src/middlewared/middlewared/plugins/virt/disk.py

```diff
@@ -0,0 +1,178 @@
+import errno
+import os
+import re
+import shlex
+import subprocess
+
+from middlewared.api import api_method
+from middlewared.api.current import (
+    VirtDeviceImportDiskImageArgs, VirtDeviceImportDiskImageResult, VirtDeviceExportDiskImageArgs,
+    VirtDeviceExportDiskImageResult,
+)
+from middlewared.plugins.zfs_.utils import zvol_name_to_path
+from middlewared.service import CallError, job, Service
+
+
+# Valid Disk Formats we can export
+VALID_DISK_FORMATS = ['qcow2', 'qed', 'raw', 'vdi', 'vpc', 'vmdk']
+
+
+class VirtDeviceService(Service):
+
+    class Config:
+        namespace = 'virt.device'
+        cli_namespace = 'virt.device'
+
+    @api_method(VirtDeviceImportDiskImageArgs, VirtDeviceImportDiskImageResult, roles=['VIRT_INSTANCE_WRITE'])
+    @job(lock_queue_size=1, lock=lambda args: f"virt_zvol_disk_image_{args[-1]['zvol']}")
+    def import_disk_image(self, job, data):
+        """
+        Imports a specified disk image.
+
+        Utilized qemu-img with the auto-detect functionality to auto-convert
+        any supported disk image format to RAW -> ZVOL
+
+        As of this implementation it supports:
+
+        - QCOW2
+        - QED
+        - RAW
+        - VDI
+        - VPC
+        - VMDK
+
+        `diskimg` is a required parameter for the incoming disk image
+        `zvol` is the required target for the imported disk image
+        """
+        if not self.middleware.call_sync('zfs.dataset.query', [('id', '=', data['zvol'])]):
+            raise CallError(f"zvol {data['zvol']} does not exist.", errno.ENOENT)
+
+        if os.path.exists(data['diskimg']) is False:
+            raise CallError('Disk Image does not exist.', errno.ENOENT)
+
+        if os.path.exists(zvol_name_to_path(data['zvol'])) is False:
```

Comment:
You're calling `zvol_name_to_path` twice unnecessarily. Let's call it once and then do checks.